src/kv/mod.rs

```diff
@@ -1,35 +1,155 @@
 
 /// Key-Value client.
 #[derive(Clone)]
 pub struct Kv {
     /// Etcd Key-Value client.
     client: KvClient,
+    /// Etcd client cache.
+    cache: Cache,
+    /// Enable etcd client cache.
+    cache_enable: bool,
+    /// Etcd watch request sender.
+    watch_sender: Sender<EtcdWatchRequest>,
 }
 
+/// Etcd client cache default size.
+const ETCD_CACHE_DEFAULT_SIZE: usize = 64;
+
 impl Kv {
     /// Creates a new `KvClient`.
-    pub(crate) const fn new(client: KvClient) -> Self {
-        Self { client }
+    ///
+    /// This method should only be called within etcd client.
+    pub(crate) fn new(
+        client: KvClient,
+        watch_client: &WatchClient,
+        cache_size: usize,
+        cache_enable: bool,
+    ) -> Self {
+        let etcd_cache_size = if cache_size == 0 {
+            ETCD_CACHE_DEFAULT_SIZE
+        } else {
+            cache_size
+        };
+
+        let (watch_req_sender, watch_req_receiver) = unbounded::<EtcdWatchRequest>();
+
+        let cache = Cache::new(etcd_cache_size, watch_req_sender.clone());
+        let (mut client_req_sender, mut client_resp_receiver) = watch_client
+            .watch()
+            .unwrap_or_else(|e| panic!("failed to send watch commend, the response is: {}", e));
+
+        let cache_inner = cache.clone();
+        let mut processing_key = vec![];
+        let mut processing_watch_req = false;
+        smol::spawn(async move {
+            loop {
+                if !processing_watch_req {
+                    if let Ok(watch_req) = watch_req_receiver.try_recv() {
+                        processing_key = watch_req.get_key();
+                        client_req_sender
+                            .send((watch_req.into(), WriteFlags::default()))
+                            .await
+                            .unwrap_or_else(|e| {
+                                panic!("Fail to send watch request, the error is: {}", e)
+                            });
+                        processing_watch_req = true;
+                    }
+                }
+                match client_resp_receiver.next().await {
+                    Some(watch_resp) => {
+                        match watch_resp {
+                            Ok(resp) => {
+                                if resp.get_created() {
+                                    cache_inner.insert_watch_id(
+                                        processing_key.clone(),
+                                        resp.get_watch_id(),
+                                    );
+                                    processing_watch_req = false;
+                                    continue;
+                                }
+                                if resp.get_canceled() {
+                                    cache_inner.delete_watch_id(&processing_key.clone());
+                                    processing_watch_req = false;
+                                    continue;
+                                }
```

Comment:
Do we need to check processing_watch_req is true for these two cases?

Reply:
Probably not, since only when processing_watch_req is true, we can get the watch create/cancel response